ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll

```diff
@@ -8,23 +8,236 @@ private import codeql.ruby.Concepts
+          // ActiveStorage::Blob.compose(blobs : [Blob]) : Blob
+          "activestorage;Blob;activestorage;;Member[ActiveStorage].Member[Blob].Method[compose].ReturnValue",
+          // gives error: Invalid name 'Element' in access path
+          // "activestorage;Blob;activestorage;;Member[ActiveStorage].Member[Blob].Method[compose].Argument[0].Element[any]",
+          // ActiveStorage::Blob.find_signed(!) : Blob
+          "activestorage;Blob;activestorage;;Member[ActiveStorage].Member[Blob].Method[find_signed,find_signed!].ReturnValue",
+          // ActiveStorage::Attachment#blob : Blob
+          "activestorage;Blob;activestorage;;Member[ActiveStorage].Member[Attachment].Instance.Method[blob].ReturnValue",
+          // ActiveStorage::Attachment delegates method calls to its associated Blob
+          "activestorage;Blob;activestorage;;Member[ActiveStorage].Member[Attachment].Instance",
+        ]
+    }
+  }
+
+  /**
+   * Method calls on `ActiveStorage::Blob` that send HTTP requests.
+   */
+  private class BlobRequestCall extends HTTP::Client::Request::Range {
+    BlobRequestCall() {
+      this =
+        [
+          // Class methods
+          API::getTopLevelMember("ActiveStorage")
+              .getMember("Blob")
+              .getASubclass*()
+              .getAMethodCall(["create_after_unfurling!", "create_and_upload!"]),
+          // Instance methods
+          ModelOutput::getATypeNode("activestorage", "Blob")
+              .getAMethodCall([
+                  "upload", "upload_without_unfurling", "download", "download_chunk", "delete",
+                  "purge"
+                ])
+        ].asExpr().getExpr()
+    }
+
+    override string getFramework() { result = "activestorage" }
+
+    override DataFlow::Node getResponseBody() { result.asExpr().getExpr() = this }
+
+    override DataFlow::Node getAUrlPart() { none() }
+
+    override predicate disablesCertificateValidation(DataFlow::Node disablingNode) { none() }
+  }
+
+  /**
+   * A call to `has_one_attached` or `has_many_attached`, which declares an
+   * association between an ActiveRecord model and an ActiveStorage attachment.
+   *
+   * ```rb
+   * class User < ActiveRecord::Base
+   *   has_one_attached :avatar
+   * end
+   * ```
+   */
+  private class Association extends ActiveRecordAssociation {
+    Association() { this.getMethodName() = ["has_one_attached", "has_many_attached"] }
+  }
+
+  /**
+   * An ActiveStorage attachment, instantiated via an association with an
+   * ActiveRecord model.
+   *
+   * ```rb
+   * class User < ActiveRecord::Base
+   *   has_one_attached :avatar
+   * end
+   *
+   * user = User.find(id)
+   * user.avatar
+   * ```
+   */
+  private class AttachmentInstance extends DataFlow::CallNode {
+    AttachmentInstance() {
+      exists(Association assoc, string model | model = assoc.getTargetModelName() |
+        this.getReceiver().(ActiveRecordInstance).getClass() = assoc.getSourceClass() and
+        (
```

Comment:
Are there associations that are neither singular or collections? Otherwise this looks equivalent to just `this.getMethodName() = model`.